File: java/addTwoHugeNumbers.java
/*
You're given 2 huge integers represented by linked lists. Each linked list element is a number from 0 to 9999 that represents a number with exactly 4 digits. The represented number might have leading zeros. Your task is to add up these huge integers and return the result in the same format.
*/

ListNode<Integer> addTwoHugeNumbers(ListNode<Integer> a, ListNode<Integer> b) {
    if(a == null){
        return b;
    }
    if(b == null){
        return a;
    }
    int sa = size(a);
    int sb = size(b);
    for (int i = sa; i < sb; i++) {
        ListNode<Integer> temp = new ListNode<Integer>(0);
        temp.next = a;
        a = temp;
    }
    for (int i = sb; i < sa; i++) {
        ListNode<Integer> temp = new ListNode<Integer>(0);
        temp.next = b;
        b = temp;
    }
    ListNode<Integer> at = revert(a);
    ListNode<Integer> bt = revert(b);
    int remember = 0;
    ListNode<Integer> asum = at;
    ListNode<Integer> bsum = bt;
    while(asum != null && bsum != null){
        int sum = asum.value + bsum.value + remember;
        remember = sum/10000;
        sum = sum%10000;
        asum.value = sum;
        asum = asum.next;
        bsum = bsum.next;
    }
    if(remember > 0){
        asum = at;
        while(asum.next != null){
            asum = asum.next;
        }
        asum.next = new ListNode<Integer>(remember);
    }
    return revert(at);
}

ListNode<Integer> revert(ListNode<Integer> root){
    if(root == null){
        return null;
    }
    Stack<ListNode<Integer>> stack = new Stack<ListNode<Integer>>();
    ListNode<Integer>temp = root;
    while(temp != null){
        stack.add(temp);
        temp = temp.next;
    }
    ListNode<Integer> result = stack.pop();
    temp = result; 
    while(stack.size() > 0){
        if(temp.equals(result)){
            result.next = stack.pop();
            temp = result.next;
        }else{
            temp.next = stack.pop();
            temp = temp.next;
        }
    }
    temp.next = null;
    return result;
}
int size(ListNode<Integer> a){
    if(a == null){
        return 0;
    }
    int count = 0;
    while(a != null){
        a = a.next;
        count++;
    }
    return count;
}

File: golang/addTwoHugeNumbers.go
/*
You're given 2 huge integers represented by linked lists. Each linked list element is a number from 0 to 9999 that represents a number with exactly 4 digits. The represented number might have leading zeros. Your task is to add up these huge integers and return the result in the same format.
*/
func addTwoHugeNumbers(a *ListNode, b *ListNode) *ListNode {
	var lenA = count(a);
	var lenB = count(b);
    if(lenA == 0){
        return b;
    }
    if(lenB == 0){
        return a;
    }
	if(lenA > lenB){
		 return addTwoNumbers(a, b)
	}
	return addTwoNumbers(b, a)
}

func addTwoNumbers(langer *ListNode, small *ListNode) *ListNode {
	var l = revert(langer);
	var s = revert(small);
	var nodeL = l;
	var nodeS = s;
	var mem = 0;
    var pre = nodeL;
	for nodeS != nil {
		var temp = nodeL.Value.(int) + nodeS.Value.(int) + mem;
		nodeL.Value = temp%10000;
		mem = temp / 10000;
        pre = nodeL;
		nodeL = nodeL.Next;
		nodeS = nodeS.Next;
	}
	for nodeL != nil && mem > 0 {
		var temp = nodeL.Value.(int) + mem;
		nodeL.Value = temp%10000;
		mem = temp / 10000;
        pre = nodeL
		nodeL = nodeL.Next;
	}
    if(mem > 0){
        pre.Next = &ListNode{
            Next: nil,
            Value:  mem,
        }
    }
	var result = revert(l);
	return result;
}
func revert(l *ListNode) (*ListNode){
	var current = l;
	if(l == nil || l.Next == nil){
		return l;
	}
	var next = l.Next;
	if(next.Next == nil){
		next.Next = current;
		current.Next = nil;
		return next;
	}
	var nextNext = next.Next;
	current.Next = nil;
	for nextNext != nil {
		next.Next = current;
		current = next;
		next = nextNext;
		nextNext = nextNext.Next;
	}
	next.Next = current;
	return next;
}

func count(a *ListNode) (int){
	if(a == nil){
		return 0;
	}
	return count(a.Next) + 1;
}
